lib/phoenix_live_view/html_engine.ex

```diff
@@ -244,4 +244,307 @@ defmodule Phoenix.LiveView.HTMLEngine do
+         ],
+         meta,
+         {hooks, tokens_acc}
+       ) do
+    str_attrs = for {name, {:string, value, _}, _} <- attrs, into: %{}, do: {name, value}
+    %{line: line, column: column} = start_meta
+    hook_meta = Map.merge(meta, %{line: line, column: column})
+
+    case classify_hook(str_attrs) do
+      # keep runtime hooks
+      {:runtime, name} ->
+        # keep bundle="runtime" hooks in DOM
+        {hooks, start, content, end_} =
+          process_runtime_hook(hooks, name, start, content, end_, hook_meta)
+
+        process_hooks(rest, meta, {hooks, [end_, content, start | tokens_acc]})
+
+      {:bundle_current, name} ->
+        # only consider bundle="current_otp_app" hooks if they are part of the current otp_app
+        if current_otp_app() == colocated_hooks_app() do
+          IO.puts(
+            "Adding hook #{name} from colo #{colocated_hooks_app()} == current #{current_otp_app()}"
+          )
+
+          hooks = process_bundled_hook(hooks, name, text, hook_meta)
+          process_hooks(rest, meta, {hooks, tokens_acc})
+        else
+          IO.puts("Skipping hook #{name} from #{colocated_hooks_app()}")
+          process_hooks(rest, meta, {hooks, tokens_acc})
+        end
+
+      {:bundle, name} ->
+        # by default, hooks with no special bundle value are extracted, no matter where they're from
+        hooks = process_bundled_hook(hooks, name, text, hook_meta)
+        process_hooks(rest, meta, {hooks, tokens_acc})
+
+      :invalid ->
+        # TODO: nice error message
+        raise ArgumentError,
+              "scripts with type=\"text/phx-hook\" must have a compile-time string \"name\" attribute"
+
+      :no_hook ->
+        process_hooks(rest, meta, {hooks, [end_, content, start | tokens_acc]})
+    end
+  end
+
+  # if the first clause did not match (tag open, text, close),
+  # this means that there is interpolation inside the script, which is not supported
+  # for colocated hooks
+  defp process_hooks([{:tag, "script", attrs, _meta} = start | rest], meta, {hooks, tokens_acc}) do
+    if Enum.find(attrs, &match?({"type", {:string, "text/phx-hook", _}, _}, &1)) do
+      # TODO: nice error message
+      raise ArgumentError,
+            "scripts with type=\"text/phx-hook\" must not contain any interpolation!"
+    else
+      process_hooks(rest, meta, {hooks, [start | tokens_acc]})
+    end
+  end
+
+  defp process_hooks([token | rest], meta, {hooks, tokens_acc}),
+    do: process_hooks(rest, meta, {hooks, [token | tokens_acc]})
+
+  defp process_hooks([], _meta, acc), do: acc
+
+  defp classify_hook(%{"type" => "text/phx-hook", "name" => name, "bundle" => "runtime"}),
+    do: {:runtime, name}
+
+  defp classify_hook(%{"type" => "text/phx-hook", "name" => name, "bundle" => "current_otp_app"}),
+    do: {:bundle_current, name}
+
+  defp classify_hook(%{"type" => "text/phx-hook", "name" => name}), do: {:bundle, name}
+  defp classify_hook(%{"type" => "text/phx-hook"}), do: :invalid
```

Comment:
My concern here is that someone may write `type={if foo, do: "text/phx-hook"}` and then it won't work properly. I wonder if we should insert a special `:bundle="..."` attribute, which would give us more control (they are always handled specially). Perhaps it could also be something that we would use for collocated CSS.

Reply:
True. We could have a big warning in the docs though. I'm not sure about making it a special attribute (those are handled in the tag_engine) as it is quite HTML specific. That's why I currently have all the code in `Phoenix.LiveView.HTMLEngine`.